runtime/wasm/src/module/mod.rs

```diff
@@ -269,126 +123,367 @@ impl WasmiModule {
+        link!("typeConversion.bigIntToHex", big_int_to_hex, ptr);
+        link!("typeConversion.stringToH160", string_to_h160, ptr);
+        link!("typeConversion.bytesToBase58", bytes_to_base58, ptr);
+
+        link!("json.fromBytes", json_from_bytes, ptr);
+        link!("json.try_fromBytes", json_try_from_bytes, ptr);
+        link!("json.toI64", json_to_i64, ptr);
+        link!("json.toU64", json_to_u64, ptr);
+        link!("json.toF64", json_to_f64, ptr);
+        link!("json.toBigInt", json_to_big_int, ptr);
+
+        link!("crypto.keccak256", crypto_keccak_256, ptr);
+
+        link!("bigInt.plus", big_int_plus, x_ptr, y_ptr);
+        link!("bigInt.minus", big_int_minus, x_ptr, y_ptr);
+        link!("bigInt.times", big_int_times, x_ptr, y_ptr);
+        link!("bigInt.dividedBy", big_int_divided_by, x_ptr, y_ptr);
+        link!("bigInt.dividedByDecimal", big_int_divided_by_decimal, x, y);
+        link!("bigInt.mod", big_int_mod, x_ptr, y_ptr);
+        link!("bigInt.pow", big_int_pow, x_ptr, exp);
+
+        link!("bigDecimal.toString", big_decimal_to_string, ptr);
+        link!("bigDecimal.fromString", big_decimal_from_string, ptr);
+        link!("bigDecimal.plus", big_decimal_plus, x_ptr, y_ptr);
+        link!("bigDecimal.minus", big_decimal_minus, x_ptr, y_ptr);
+        link!("bigDecimal.times", big_decimal_times, x_ptr, y_ptr);
+        link!("bigDecimal.divided_by", big_decimal_divided_by, x, y);
+        link!("bigDecimal.equals", big_decimal_equals, x_ptr, y_ptr);
+
+        link!("dataSource.create", data_source_create, name, params);
+        link!(
+            "dataSource.createWithContext",
+            data_source_create_with_context,
+            name,
+            params,
+            context
         );
+        link!("dataSource.address", data_source_address,);
+        link!("dataSource.network", data_source_network,);
+        link!("dataSource.context", data_source_context,);
+
+        link!("ens.nameByHash", ens_name_by_hash, ptr);
+
+        link!("log.log", log_log, level, msg_ptr);
+
+        link!("arweave.transactionData", arweave_transaction_data, ptr);
+
+        link!("box.profile", box_profile, ptr);
+
+        let instance = linker.instantiate(&valid_module.module)?;
+
+        // Provide access to the WASM runtime linear memory
+        let memory = instance
+            .get_memory("memory")
+            .context("Failed to find memory export in the WASM module")?;
+
+        let memory_allocate = instance
+            .get_func("memory.allocate")
+            .context("`memory.allocate` function not found")?
+            .get1()?;
+
+        let should_interrupt = Arc::new(AtomicBool::new(true));
+        if let Some(timeout) = timeout {
+            // This task is likely to outlive the instance, which is fine.
+            let interrupt_handle = instance.store().interrupt_handle().unwrap();
+            let should_interrupt = should_interrupt.clone();
+            graph::spawn(async move {
```

Comment:
It is possible to construct a subgraph that intentionally avoids timeout and enters an infinite loop. There are several problems here:

1. Calling `ipfs_map` once is enough to disable the timeout forever. The timeout is never restored, so the subgraph could enter an infinite loop right after this.
2. Even if the timeout was restored, this check never triggers a second time if it happens to check while the timeout was disabled
3. The use of atomics is overkill for something that is not on the hot path
4. The code that disables `should_interrupt` is limited to `ipfs_map` but does not count eg: `ipfs_cat` which may also want to disable the inturrupt.
5. Even if the inturrupt were disabled for just the duration of the ipfs_map call and we turned it back on at the end, code spawned in other modules needs to count toward the limit.

We discussed outside of the context of this PR that we will want to handle timeouts in a more deterministic way, but before I get to writing that RFC I think we need to do better here since this is an attack vector as-is.

I suggest writing something like a shared `Arc<Mutex<Stopwatch>>` that can be paused for only the duration of time that is external to any possible wasm code executing (eg: inside ipfs reading code, but be running for ipfs_map callbacks) Or maybe just not disabling the timeout at all.